pkg/deploy/predeploy.go

```diff
@@ -376,54 +379,71 @@ func (d *deployer) configureServiceSecrets(ctx context.Context) error {
 		{d.serviceKeyvault, env.EncryptionSecretName, 32},
 		{d.serviceKeyvault, env.FrontendEncryptionSecretName, 32},
 	} {
-		err := d.ensureSecret(ctx, s.kv, s.secretName, s.len)
+		isNew, err := d.ensureSecret(ctx, s.kv, s.secretName, s.len)
+		isRotated = isNew || isRotated
 		if err != nil {
 			return err
 		}
 	}
 
-	return d.ensureSecretKey(ctx, d.portalKeyvault, env.PortalServerSSHKeySecretName)
+	isNew, err := d.ensureSecretKey(ctx, d.portalKeyvault, env.PortalServerSSHKeySecretName)
+	isRotated = isNew || isRotated
+	if err != nil {
+		return err
+	}
+
+	if isRotated {
+		err = d.restartOldScalesets(ctx, "systemctl restart aro-gateway", d.config.GatewayResourceGroupName)
+		if err != nil {
+			return err
+		}
+		err = d.restartOldScalesets(ctx, "systemctl restart aro-rp", d.config.RPResourceGroupName)
```

Comment:
IIRC aro-rp has some pretty good control sequences that'll prevent it from being killed if it's in the middle of a job, so this command could hang for as long as a cluster creation takes, right? Or has that gone away with hive? If that's true, this change then introduces a massive lag in our deployment process, where we could be adding something like 60 minutes of time for each rollout. Would it be better if we instead only restarted if we detected that the new VMSS was not coming up in an expected period of time, or would that still introduce the bug we're trying to solve?

Reply:
as a future state, would there be a way for us to instead just refresh our cosmosdb key continuously somehow with what we expect it to be as opposed to only refreshing on boot, or is that just way too hard?

Reply:
@SudoBrendan , related to aro-rp taking time for getting killed in between the job, I tested the below flow on the fix:-
- Started Cluster Creation
- Restarted aro-rp service before cluster creation completion(by redeploying the RP), took similar time as for no job running on service.
-  Cluster created successfully.

Does that answer that there won't be a lag in the deployment process?

Reply:
I had a talk with Ben about this yesterday. We came to the conclusion that `backend` processes can fail to shutdown, and that can take [60 minutes to resolve at a systemd level by design](https://github.com/Azure/ARO-RP/pull/157). This isn't going to happen each time, but it is a possibility.

Reply:
I guess for now we can just keep this in mind as something to watch out for when deploying. If we find that we're hitting this often, we can revert this change and attempt authoring continuous refreshing as part of the RP and Gateway.

Reply:
Thanks @SudoBrendan, noted. Makes sense, I will try figuring out the continuous refreshing before hand as an alternate to this PR in case the lag is frequent.